components/time/src/zone/iana.rs

```diff
@@ -548,40 +527,74 @@ impl<'a> IanaParserExtendedBorrowed<'a> {
     /// use icu::time::zone::iana::IanaParserExtended;
     /// use icu::time::zone::TimeZone;
     /// use std::collections::BTreeMap;
-    /// use tinystr::tinystr;
+    /// use icu::locale::subtags::subtag;
     ///
     /// let parser = IanaParserExtended::new();
     ///
-    /// let ids = parser.iter_all().enumerate().map(|(a, b)| (b, a)).collect::<std::collections::BTreeMap<_, _>>();
+    /// let ids = parser.iter_all().enumerate().map(|(a, b)| ((b.time_zone, b.canonical, b.normalized), a)).collect::<std::collections::BTreeMap<_, _>>();
     ///
-    /// let kyiv_idx = ids[&(TimeZone(tinystr!(8, "uaiev")), "Europe/Kyiv", "Europe/Kyiv")];
-    /// let kiev_idx = ids[&(TimeZone(tinystr!(8, "uaiev")), "Europe/Kyiv", "Europe/Kiev")];
-    /// let uzgh_idx = ids[&(TimeZone(tinystr!(8, "uaiev")), "Europe/Kyiv", "Europe/Uzhgorod")];
-    /// let zapo_idx = ids[&(TimeZone(tinystr!(8, "uaiev")), "Europe/Kyiv", "Europe/Zaporozhye")];
+    /// let kyiv_idx = ids[&(TimeZone(subtag!("uaiev")), "Europe/Kyiv", "Europe/Kyiv")];
+    /// let kiev_idx = ids[&(TimeZone(subtag!("uaiev")), "Europe/Kyiv", "Europe/Kiev")];
+    /// let uzgh_idx = ids[&(TimeZone(subtag!("uaiev")), "Europe/Kyiv", "Europe/Uzhgorod")];
+    /// let zapo_idx = ids[&(TimeZone(subtag!("uaiev")), "Europe/Kyiv", "Europe/Zaporozhye")];
     ///
     /// // The order for a particular time zone is guaranteed
     /// assert!(kyiv_idx < kiev_idx && kiev_idx < uzgh_idx && uzgh_idx < zapo_idx);
     /// // It is not guaranteed that the entries for a particular time zone are consecutive
     /// assert!(kyiv_idx + 1 != kiev_idx);
     ///
-    /// assert_eq!(ids.len(), 598);
+    /// assert!(parser.iter_all().count() >= 598);
     /// ```
     pub fn iter_all(&self) -> TimeZoneAndCanonicalAndNormalizedIter<'a> {
         TimeZoneAndCanonicalAndNormalizedIter(0, *self)
     }
 }
 
+/// Return value of [`IanaParserBorrowed::iter`].
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct TimeZoneAndCanonical<'a> {
+    /// The parsed [`TimeZone`]
+    pub time_zone: TimeZone,
+    /// The canonical IANA ID
+    pub canonical: &'a str,
+}
+
+/// Return value of [`IanaParserExtendedBorrowed::parse`], [`IanaParserExtendedBorrowed::iter`].
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct TimeZoneAndCanonicalAndNormalized<'a> {
```

Comment:
Thought: Not excited about these names but I don't have anything better to suggest